### PR TITLE
Add super resolution example (ESPCN).

### PR DESCRIPTION
**Pull request by xingyu-long**
@fchollet Please review this examples. Thanks.

### CODE REVIEW

File: examples/vision/super_resolution_sub_pixel.py [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+            logits = model(x_batch_train, training=True)
+            loss_value = loss_fn(y_batch_train, logits)
+        grads = tape.gradient(loss_value, model.trainable_weights)
+        optimizer.apply_gradients(zip(grads, model.trainable_weights))
+        total_loss += loss_value
+        print(
+            "Training loss (for one batch) at step %d: %.4f" % (step, float(loss_value))
+        )
+    print(
+        "Avg. Loss for each epoch %d: %.4f" % (epoch + 1, total_loss / len(train_gen))
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:374]: You can replace all of this with `model.fit()`, much nicer (and feature-complete, since you can now train on multiple GPUs or TPUs with no code change)

File: examples/vision/super_resolution_sub_pixel.py [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+    )
+    # Run a validation loop at the end of each epoch.
+    for x_batch_val, y_batch_val in valid_gen:
+        val_logits = model(x_batch_val, training=False)
+        val_loss_val = loss_fn(y_batch_val, val_logits)
+        # Calculate PSNR value.
+        psnr = 10 * math.log10(1 / val_loss_val)
+        total_psnr += psnr
+        validation_loss += val_loss_val
+    print("Avg. PSNR for each epoch: %.4f" % (total_psnr / len(valid_gen)))
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:384]: You can compute average PSNR per epoch in a callback (which would print the result at the end of each epoch). See: https://keras.io/guides/writing_your_own_callbacks/

File: examples/vision/super_resolution_sub_pixel.py [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+channels = 1
+
+batch_size = 32
+
+# input_shape = (input_img_size[0], input_img_size[1], channels)
+
+from tensorflow.keras.layers import Conv2D
+
+
+class ESPCN(tf.keras.Model):
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:238]: Please use the functional API to define this model, it will be less verbose, more readable, and less error-prone.
```python
conv_args = {"activation": "relu", "kernel_initializer": "orthogonal", "padding": "same"}
inputs = keras.Inputs(shape=input_img_size + (1,))
x = layers.Conv2D(64, 5, **conv_args)(inputs)
x = layers.Conv2D(64, 3, **conv_args)(x)
x = layers.Conv2D(64, 3, **conv_args)(x)
x = layers.Conv2D(channels * upscale_factor ** 2, 3, **conv_args)(x)
outputs = tf.nn.depth_to_space(x, upscale_factor)
model = keras.Model(inputs, outputs)
model.summary()
```

Did you really mean `channels * upscale_factor ** 2`, or `(channels * upscale_factor) ** 2`?

File: examples/vision/super_resolution_sub_pixel.py [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+"""
+## Crop and resize images.
+
+Since we need to upscale images and we have to process the input data format. First, you
+need to define crop size and calculate the acutal crop size by following functions. For
+input data, after we get the crop image and blur it with `bicubic` method. For target
+data, we only do the crop operation.
+"""
+
+import PIL
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:103]: Rather than doing everything in PIL, which won't scale very well and takes up a lot of code, prefer creating a tf.data.Dataset (e.g. with `image_dataset_from_directory()` and process images with TF ops.

You can use https://www.tensorflow.org/api_docs/python/tf/image/rgb_to_yuv and its inverse yuv_to_rgb

Probably something like:

```python
def process_input(x):
   return tf.image.resize(x, ...)

def process_target(x):
   return tf.image.crop(x, ...)

dataset = image_dataset_from_directory(directory, label_mode=None)
dataset = dataset.map(rgb_to_yuv)
dataset = dataset.map(lambda x: (process_input(x), process_target(x)))
```

File: examples/vision/super_resolution_sub_pixel.py [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+
+import matplotlib.pyplot as plt
+import numpy as np
+import math
+
+from mpl_toolkits.axes_grid1.inset_locator import zoomed_inset_axes
+from mpl_toolkits.axes_grid1.inset_locator import mark_inset
+
+
+def plot_results(img, title):
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:416]: You could also put this logic in a callback, and save an image of the results to disk at the end of each epoch. It's great to see how the model progresses during training

File: examples/vision/super_resolution_sub_pixel.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    plt.yticks(visible=False)
+    plt.xticks(visible=False)
+
+    # Make the line.
+    mark_inset(ax, axins, loc1=1, loc2=3, fc="none", ec="blue")
+
+    plt.show()
+
+
+def predict_result(model, lr_img):
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:441]: What's this? Please add docstrings to these functions.

File: examples/vision/super_resolution_sub_pixel.py (hunk 3 above)
Comment (xingyu-long) [reply examples/vision/super_resolution_sub_pixel.py:238]: It should be `channels * (upscale_factor ** 2)`.

File: examples/vision/super_resolution_sub_pixel.py [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+    print("Avg. PSNR for each epoch: %.4f" % (total_psnr / len(valid_gen)))
+    print(
+        "Avg. Validation Loss for each epoch: %.4f" % (validation_loss / len(valid_gen))
+    )
+    best_psnr = max(best_psnr, float(total_psnr / len(valid_gen)))
+
+print("Best PSNR: %.4f" % best_psnr)
+
+"""
+Save model by `model.save()`
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:393]: nit: Save model by `model.save()`. By default, it's the TensorFlow SavedModel format.

File: examples/vision/super_resolution_sub_pixel.py [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+"""
+
+"""
+## Introduction
+
+ESPCN (Efficient Sub-Pixel CNN), proposed by [Shi, W.
+2016](https://arxiv.org/abs/1609.05158) is used to reconstruct the image from low
+resolution (LR) to high resolution (HR). In this paper, they introduce an efficient
+sub-pixel convolution layers which learns an array of upscaluing filter to upscale the
+final LR feature maps into the HR output. This post will implement the model in this
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:16]: nit: update LR and HR here.

File: examples/vision/super_resolution_sub_pixel.py [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+paper and trained with relatively small dataset
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:17]: nit: paper and train it with a small dataset.

File: examples/vision/super_resolution_sub_pixel.py [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+"""
+## Setup
+"""
+
+import tensorflow as tf
+
+"""
+## Load data: BSDS500 dataset
+
+### Download data
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:30]: nit: dataset

File: examples/vision/super_resolution_sub_pixel.py [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+        if fname.endswith(".jpg")
+    ]
+)
+
+print("Number of train samples: ", len(input_img_paths))
+print("Number of test samples: ", len(test_img_paths))
+print("Number of validation samples: ", len(valid_img_paths))
+
+"""
+## Visualiza the data
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:86]: nit: Visualize

File: examples/vision/super_resolution_sub_pixel.py (hunk 4 above)
Comment (yhliang2018) [reply examples/vision/super_resolution_sub_pixel.py:103]: +1 on this! `tf.data.Dataset` is the recommended API for the input pipeline of TF 2.x. This will be a good showcase of it.

File: examples/vision/super_resolution_sub_pixel.py [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+"""
+
+from IPython.display import display
+from tensorflow.keras.preprocessing.image import load_img
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:90]: This import is not necessary with the `tf` import.

File: examples/vision/super_resolution_sub_pixel.py [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+
+display(load_img(input_img_paths[0]))
+
+"""
+## Crop and resize images.
+
+Since we need to upscale images and we have to process the input data format. First, you
+need to define crop size and calculate the acutal crop size by following functions. For
+input data, after we get the crop image and blur it with `bicubic` method. For target
+data, we only do the crop operation.
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:100]: nit: Can you rephrase this part as something similar to the following?
" Let's process the image data format and upscale the images. For the input data, we the crop image and blur it with `bicubic` method; For the target data, we only do the crop operation."

File: examples/vision/super_resolution_sub_pixel.py [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+"""
+
+"""
+We luminance in `YCbCr` colour space for images because the changes in that space are
+easy to see.
+"""
+
+from tensorflow.keras.preprocessing.image import array_to_img
+from tensorflow.keras.preprocessing.image import img_to_array
+import numpy as np
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:132]: Let's use `tf.xxx` APIs directly if possible, rather than importing the symbols directly.

File: examples/vision/super_resolution_sub_pixel.py [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+input_img_size = (crop_size // upscale_factor, crop_size // upscale_factor)
+target_img_size = (crop_size, crop_size)
+
+channels = 1
+
+batch_size = 32
+
+# input_shape = (input_img_size[0], input_img_size[1], channels)
+
+from tensorflow.keras.layers import Conv2D
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:235]: Ditto as above.

File: examples/vision/super_resolution_sub_pixel.py [hunk 16]
```diff
@@ -0,0 +1,513 @@
...
+
+
+def get_model(upscale_factor=3, channels=1):
+
+    model = ESPCN(channels, upscale_factor)
+    return model
+
+
+"""
+ESPCN class create same model with following code.
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:274]: nit: creates the same model with the following code.

File: examples/vision/super_resolution_sub_pixel.py [hunk 17]
```diff
@@ -0,0 +1,513 @@
...
+    input_img_paths=valid_img_paths,
+    transform_input=transform_input,
+    transform_target=transform_target,
+    batch_size=batch_size,
+    upscale_factor=upscale_factor,
+    channels=channels,
+)
+
+"""
+## Train the model with custom training loop.
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:344]: nit: and perform evaluation.

File: examples/vision/super_resolution_sub_pixel.py [hunk 18]
```diff
@@ -0,0 +1,513 @@
...
+"""
+Save model by `model.save()`
+"""
+
+model_path = os.path.join(os.getcwd(), "models")
+
+model.save(model_path)
+
+"""
+## Predict and plot the results.
```
Comment (yhliang2018) [inline examples/vision/super_resolution_sub_pixel.py:401]: nit: Run model prediction and .....

File: examples/vision/super_resolution_sub_pixel.py (hunk 14 above)
Comment (xingyu-long) [reply examples/vision/super_resolution_sub_pixel.py:132]: I checked other examples on keras.io, they didn't use in that way, they usually import like this.
```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```
I think I should keep the consistency.

File: examples/vision/super_resolution_sub_pixel.py (hunk 5 above)
Comment (xingyu-long) [reply examples/vision/super_resolution_sub_pixel.py:416]: How do we get image input in callback if I use the `fit` ?  I didn't fin it on keras.io.

**Comment by xingyu-long** [comment]
> Do you have some visualizations of the results?

You can check with [colab](https://colab.research.google.com/drive/1Fbgqz-P5WoVh7ozd_LQZCGsDxE-_ejkL?usp=sharing) for now. I also save the results after we predict.

**Comment by xingyu-long** [comment]
@fchollet @yhliang2018  For now, after predict the results we need to merge from `y, cb, cr` channels, I followed this [examples](https://github.com/leftthomas/ESPCN/blob/master/test_image.py#L45) And I compared rgb_to_yuv and image.convert('YCbCr') in this colab file and it shows different values and even couldn't restore the img. You can check from [here](https://colab.research.google.com/drive/1zsrT2rKTEd_x4RA9geyD524fIXg4IOb2?usp=sharing). So, for me, I prefer to stay with PIL image.

File: examples/vision/super_resolution_sub_pixel.py (hunk 5 above)
Comment (fchollet) [reply examples/vision/super_resolution_sub_pixel.py]: Just pass a few image paths as constructor arguments to your callback.

File: examples/vision/super_resolution_sub_pixel.py (hunk 14 above)
Comment (fchollet) [reply examples/vision/super_resolution_sub_pixel.py]: +1

File: examples/vision/super_resolution_sub_pixel.py [hunk 19]
```diff
@@ -0,0 +1,471 @@
+"""
+Title: Image Super-Resolution Using an Efficient Sub-Pixel CNN
+Author: [Xingyu Long](https://github.com/xingyu-long)
+Date created: 2020/07/28
+Last modified: 2020/07/30
+Description: Implementing Super-Resoluion using Efficient sub-pixel model on BSDS500.
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:6]: Resolution

File: examples/vision/super_resolution_sub_pixel.py [hunk 20]
```diff
@@ -0,0 +1,471 @@
...
+upscaling filter to upscale the final lowres feature maps into the highres output. This post will
+implement the model in this paper and train it with small dataset
+[BSDS500](http://www.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/BSR/BSR_bsds500.tgz).
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:24]: There are some imports throughout the notebook, prefer putting them all at the start of the file

File: examples/vision/super_resolution_sub_pixel.py [hunk 21]
```diff
@@ -0,0 +1,471 @@
...
+    upscale_factor=upscale_factor,
+    channels=channels,
+)
+
+"""
+## Train the model with custom callback and perform evaluation
+"""
+
+"""
+### Define callback
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:285]: "Define callbacks to monitor training"

File: examples/vision/super_resolution_sub_pixel.py [hunk 22]
```diff
@@ -0,0 +1,471 @@
...
+            "...Evaluating: end of batch {} and loss is {:7.4f}".format(
+                batch, logs["loss"]
+            )
+        )
+
+
+model = get_model(upscale_factor=upscale_factor, channels=channels)
+model.summary()
+
+callbacks = [ESPCNCallback()]
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:322]: You probably also need an EarlyStopping callback, no?

File: examples/vision/super_resolution_sub_pixel.py [hunk 23]
```diff
@@ -0,0 +1,471 @@
...
+model.compile(
+    optimizer=optimizer, loss=loss_fn,
+)
+
+model.fit(
+    train_gen, epochs=epochs, validation_data=valid_gen, callbacks=callbacks, verbose=0
+)
+
+"""
+Save model by `model.save()`.
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:341]: You could also use a ModelCheckpoint callback

**Comment by xingyu-long** [comment]
Already update it! @fchollet I tried with `tf.data` but the resizing of TF ops (I also tried with other methods when we resize the img) is not smooth compared with PIL ops and the results are bad. So, I stay with PIL ops. And the rgb_to_yuv and yuv_to_rgb, I made the mistakes before and both work fine if we put it into required range.

**Comment by fchollet** [comment]
> but the resizing of TF ops (I also tried with other methods when we resize the img) is not smooth compared with PIL ops and the results are bad

The interpolation method used for resizing is configurable. If you use the same interpolation you should get the same results (since it's the exact same algorithm being used)

**Comment by xingyu-long** [comment]
Hi @fchollet, You can check the update and already use `tf.data` for the data pipeline.

File: examples/vision/super_resolution_sub_pixel.py [hunk 24]
```diff
@@ -0,0 +1,414 @@
...
+        print(
+            "End epoch {} of training and loss is {:7.4f} "
+            "and the average PSNR is {:7.4f}".format(
+                epoch + 1, logs["loss"], sum(self.psnr) / len(self.psnr)
+            )
+        )
+        _, _, predict = get_lr_hr_predict(self.model, upscale_factor, self.test_img)
+        plot_results(predict, "epoch-" + str(epoch), "prediction")
+
+    def on_train_batch_end(self, batch, logs=None):
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:322]: Printing any metric computed on a single batch is not useful, because it's not accurate.

What you should print is the metric average over the epoch.

For the loss, the progress bar already does this.

For `psnr`, you can use a `keras.metrics.Mean` metric tracker, and do:

- in __init__: `self.psnr_tracker = keras.metrics.Mean('psnr')`
- in on_batch_end: `self.psnr_tracker.update_state(psnr)`
- in on_epoch_end:  `print('Mean PSNR for epoch: %.2f' % (self.psnr_tracker.result(),)` and `self.psnr_tracker.reset_states()`

File: examples/vision/super_resolution_sub_pixel.py [hunk 25]
```diff
@@ -0,0 +1,414 @@
...
+"""
+Let's predict a few images and save the results. <br> The dataset they use to train in
+paper is ImageNet, you can try it and it should give better performance.
+"""
+
+total_bicubic_psnr = 0.0
+total_test_psnr = 0.0
+
+for index, test_img_path in enumerate(test_img_paths[50:60]):
+    lowres_img, highres_img, predict = get_lr_hr_predict(
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:393]: This will be clearer with 3 separate functions

File: examples/vision/super_resolution_sub_pixel.py [hunk 26]
```diff
@@ -0,0 +1,414 @@
...
+"""
+
+epochs = 50
+
+model.compile(
+    optimizer=optimizer, loss=loss_fn,
+)
+
+model.fit(
+    train_ds, epochs=epochs, callbacks=callbacks, validation_data=valid_ds, verbose=0
```
[the remark this answers is not in the document]
Comment (xingyu-long) [reply examples/vision/super_resolution_sub_pixel.py:374]: Sounds great. Thanks!

File: examples/vision/super_resolution_sub_pixel.py [hunk 27]
```diff
@@ -0,0 +1,414 @@
...
+    out_img_cb = cb.resize(out_img_y.size, PIL.Image.BICUBIC)
+    out_img_cr = cr.resize(out_img_y.size, PIL.Image.BICUBIC)
+    out_img = PIL.Image.merge("YCbCr", (out_img_y, out_img_cb, out_img_cr)).convert(
+        "RGB"
+    )
+
+    return out_img
+
+
+def get_lr_hr_predict(model, upscale_factor, img_path):
```
[the remark this answers is not in the document]
Comment (xingyu-long) [reply examples/vision/super_resolution_sub_pixel.py:278]: Great idea! I will update it.

File: examples/vision/super_resolution_sub_pixel.py [hunk 28]
```diff
@@ -0,0 +1,408 @@
...
+    out_img_cb = cb.resize(out_img_y.size, PIL.Image.BICUBIC)
+    out_img_cr = cr.resize(out_img_y.size, PIL.Image.BICUBIC)
+    out_img = PIL.Image.merge("YCbCr", (out_img_y, out_img_cb, out_img_cr)).convert(
+        "RGB"
+    )
+
+    return out_img
+
+
+def get_resized_image(img, w, h):
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:286]: I expected this function to start from an image path -- if it already takes an image, then you don't need this function, just use `img.resize((w, h))` directly. You can remove this.

File: examples/vision/super_resolution_sub_pixel.py [hunk 29]
```diff
@@ -0,0 +1,408 @@
...
+        super(ESPCNCallback, self).__init__()
+        self.psnr_tracker = keras.metrics.Mean("psnr")
+        self.test_img = get_lowres_image(load_img(test_img_paths[0]), upscale_factor)
+
+    # Store PSNR value in each epoch.
+    def on_epoch_begin(self, epoch, logs=None):
+        self.psnr = []
+
+    def on_epoch_end(self, epoch, logs=None):
+        self.psnr_tracker.update_state(self.psnr)
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:320]: You're using the Mean metric as if it were just a call to `np.mean()`. If you want to do that, then you don't need the mean metric at all, you can just use `np.mean()`.

Otherwise, you'd call update_state on the current scalar value of psnr in `on_batch_end`, and query results (and reset state) in on_epoch_end.

Comment (xingyu-long) [reply examples/vision/super_resolution_sub_pixel.py:320]: Yeah, I think I can keep `np.mean`.

File: examples/vision/super_resolution_sub_pixel.py [hunk 30]
```diff
@@ -0,0 +1,399 @@
...
+    monitor="loss",
+    mode="min",
+    save_best_only=True,
+)
+
+model = get_model(upscale_factor=upscale_factor, channels=channels)
+model.summary()
+
+callbacks = [ESPCNCallback(), early_stopping_callback, model_checkpoint_callback]
+loss_fn = tf.keras.losses.MeanSquaredError()
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:341]: You've already imported Keras, so here and on the line below you can use `keras` instead of `tf.keras`

File: examples/vision/super_resolution_sub_pixel.py [hunk 31]
```diff
@@ -0,0 +1,399 @@
...
+    plt.xticks(visible=False)
+
+    # Make the line.
+    mark_inset(ax, axins, loc1=1, loc2=3, fc="none", ec="blue")
+    plt.savefig(str(prefix) + "-" + title + ".png")
+    plt.show()
+
+
+"""
+Build `upscale_image` and `get_lowres_image` method to process image.
```
Comment (fchollet) [inline examples/vision/super_resolution_sub_pixel.py:256]: Please add a couple sentences providing more details about what each function does.

**Comment by xingyu-long** [comment]
Thanks a lot!

**Comment by xingyu-long** [comment]
Hi @fchollet, I already upload these autogenerated files (it's quite a few images), thanks for your suggestions and help during this PR. Thank you!

**Comment by fchollet** [comment]
I noticed the following warning: "WARNING: Logging before flag parsing goes to stderr.\nW0828 09:22:22.304250 4660481472 functional.py:587] Model was constructed with shape (None, 100, 100, 1) for input Tensor("input_1:0", shape=(None, 100, 100, 1), dtype=float32), but it was called on an input with incompatible shape (None, 107, 160, 1)"

Have you looked into what's causing it? It seems like the data passed to the model doesn't match the shape provided in the `Input` object.

**Comment by fchollet** [comment]
Note that this warning will become an error in the next version, so we must imperatively fix it.

**Comment by fchollet** [comment]
It turns out to be an easy fix. The test image was not resized to the shape expected by the model.

The best way to fix is to specify the correct shape expectations in the input: `(None, None, 1)`.

**Comment by fchollet** [comment]
As a side note, I believe you could easily train for 10x more epochs and get better results.

**Comment by xingyu-long** [comment]
> I noticed the following warning: "WARNING: Logging before flag parsing goes to stderr.\nW0828 09:22:22.304250 4660481472 functional.py:587] Model was constructed with shape (None, 100, 100, 1) for input Tensor("input_1:0", shape=(None, 100, 100, 1), dtype=float32), but it was called on an input with incompatible shape (None, 107, 160, 1)"
> 
> Have you looked into what's causing it? It seems like the data passed to the model doesn't match the shape provided in the `Input` object.

Actually, this model should handle the arbitrary size and we only need to focus the `upscale_factor`.

**Comment by xingyu-long** [comment]
> > I noticed the following warning: "WARNING: Logging before flag parsing goes to stderr.\nW0828 09:22:22.304250 4660481472 functional.py:587] Model was constructed with shape (None, 100, 100, 1) for input Tensor("input_1:0", shape=(None, 100, 100, 1), dtype=float32), but it was called on an input with incompatible shape (None, 107, 160, 1)"
> > Have you looked into what's causing it? It seems like the data passed to the model doesn't match the shape provided in the `Input` object.
> 
> Actually, this model should handle the arbitrary size and we only need to focus the `upscale_factor`. Yeah, I can specific input as `(None, None, 1)`

**Comment by xingyu-long** [comment]
> It turns out to be an easy fix. The test image was not resized to the shape expected by the model.
> 
> The best way to fix is to specify the correct shape expectations in the input: `(None, None, 1)`.

Yeah, I see, I will do that.

**Comment by xingyu-long** [comment]
Reference: from the paper, they used ImageNet to train in 100 epochs.

**Comment by xingyu-long** [comment]
Oh, I see you change the file directly. Thanks a lot and you may need to add your name into author and `Last modified` date. Thanks for your help!
